### PR TITLE
[DO NOT MERGE] Default to XDG Base Directory conforming configuration path for new installations

### PR DESCRIPTION
**Pull request by alexander255**
This changes the default configuration directory from `~/.electron-cash` to `${XDG_CONFIG_HOME:-${HOME}/.config}/electron-cash` for new installations. Existing installations will continue to use the old directory.

This isn't directly required for the Flatpak I'm currently working on (see #625), but it ensures that the expected directory for app configuration files on Linux/Unix platforms is used. I have tested the code both sandboxed (in Flatpak) and non-sandboxed and it works in both environments and the configuration is, thanks to recent changes to the Flatpak, also shared between both versions.

Also it looks the modified piece of source code is used to determine the default configuration directory  on macOS as well? If so, then neither the previous nor the new default value (after this patch) is correct, as it should default to `~/Library/Application Support/<bundle-id>` there afaik ([Source](https://developer.apple.com/library/content/documentation/General/Conceptual/MOSXAppProgrammingGuide/AppRuntime/AppRuntime.html#//apple_ref/doc/uid/TP40010543-CH2-SW13)).

### CODE REVIEW

**Comment by marceloneil** [comment]
Currently the config _and_ data files are stored in `~/.electron-cash`. Would it make sense to split them up between `$XDG_CONFIG_HOME` and ` $XDG_DATA_HOME` rather than move them all to `$XDG_CONFIG_HOME`.

Additionally, do all posix systems use XDG? Is using the old directory rather than migrating the best method?

File: lib/util.py [hunk 1]
```diff
@@ -355,7 +355,17 @@ def user_dir():
...
         return android_check_data_dir()
     elif os.name == 'posix':
-        return os.path.join(os.environ["HOME"], ".electron-cash" )
+        # Prefer XDG Base directory spec based condiguration path, but use the legacy dot-directory
+        # configuration path if user has upgraded from a previous installation
+        xdg_config_home = os.environ.get("XDG_CONFIG_HOME", os.path.join(os.environ["HOME"], ".config"))
+        config_path_modern = os.path.join(xdg_config_home, "electron-cash")
+        config_path_legacy = os.path.join(os.environ["HOME"], ".electron-cash")
+        if os.path.exists(config_path_modern) and os.listdir(config_path_modern):
+            return config_path_modern
```
[the remark this answers is not in the document]
Comment (alexander255) [reply lib/util.py]: > Is this trying to handle a case where the user has both ~/.electron-cash and ~/.config/electron-cash?

Yes, it first probe for the new directory and if that doesn't exist (or is empty), it will probe for the old directory. If neither exist will will finally default to the new one. This is a pragmatic way to deal with the different possible locations that is used by several other software projects as well.

The alternative would be, of course, to always use the new directory and, if that isn't initialized yet, try to move the data in from the old location. This would have to be handled like the current *wallet upgrade* migrations through IMHO, since it breaks older versions after being completed. If you prefer it this way I'd be very thankful if you could provide some pointers as to were this should be implemented in the source code.

**Comment by alexander255** [comment]
> Currently the config *and* data files are stored in `~/.electron-cash`. Would it make sense to split them up between `$XDG_CONFIG_HOME` and `$XDG_DATA_HOME` rather than move them all to `$XDG_CONFIG_HOME`.

True that! I guess `wallets` & `config` would be `$XDG_CONFIG_HOME`, the rest `$XDG_DATA_HOME`? I've seen some projects being hasitant about having “more than one save location”, but it appears to me that it would split quite nicely in this case? If we do this we'll definitely need that migration through.

> Additionally, do all posix systems use XDG?

Linux definitely yes.
*BSD shares most of its userland software with Linux, so they tend to inherit things like this (at least that is my impression after some searching, I've never come accross any software having special rules for *BSD regarding where per-user data files are stored).
macOS has its own system for this (see last paragraph of the OP): It appears to me that the split between `~/Library/<app-id>` & `~/Library/Application Support/<app-id>` is similar to the split between `$XDG_CONFIG_HOME` & `$XDG_DATA_HOME`, but I'm not 100% sure about this.

And while we're already at it disecting the app dir: What does the current `~/.electrum-cash/cache` directory do? Would it be a good match for `$XDG_CACHE_HOME` or no?

**Comment by rt121212121** [comment]
EC is not my project.  You need to join the slack channel and try and
discuss it.

On Mon, May 28, 2018 at 1:20 AM, Alexander Schlarb <notifications@github.com

**Comment by alexander255** [comment]
@marceloneil: I answered your questions 2 months ago! Now please answer my 2 questions please:

  1. Do you want me to attempt to do a config/data/cache file split whose patches would be much more invasive than the current patch set? (And if not, do you want me to attempt to move all data to the new location?)
  2. Do you want me to do it only on Linux+BSD or also on macOS?

I cannot (and will not) work on stuff that requires some level of interactive discussion anymore when I have several months of latency for each not totally obvious question. This is just dumb, I'm sorry to say this.

**Comment by rt121212121** [comment]
Why won't you join the slack channel if low latency is important to you?  I told you 18 days ago you needed to join it to get the answers you want.  None of us are paid to answer these  issues, and we're all busy.  Even our own issues and pull requests don't have people to review them.

If you send me your email address to roger.taylor.email@gmail.com and I will send you a slack invite.

**Comment by marceloneil** [comment]
Hey @alexander255, sorry for the long waits. I've personally been very busy with IRL stuff so I've had to take a break from EC, only working on the essentials.

I *personally* feel no strong objection to moving the default config file or splitting the cache and config, however I wouldn't be able to merge such a change. It have to be approved by @kyuupichan first.

I also apologize for not following up on the issues I created on your gitlab repo, I'll be looking into some more thorough testing extremely soon once I've finished my exams.

Re @rt121212121: I've already offered an invite and he doesn't want to join the slack - which I don't really blame him for. Joining an invite-only proprietary chat group *just* to argue over a few PRs seems unnecessary.

**Comment by kyuupichan** [comment]
I don't have a strong opinion on this; happy to defer to others

**Comment by marceloneil** [comment]
I'm happy with this PR, my only worry is I have no clue what the expected behaviour is for non-linux posix systems. Will it work well on osx/bsd?

**Comment by fyookball** [comment]
@alexander255 can you summarize the issue for me?  As a linux user, I don't understand the benefit of splitting up the files.  Why not keep everything in ~/.electron-cash.  Splitting it up would be confusing.

**Comment by marceloneil** [comment]
Just to be clear, this PR isn't about splitting the folder (my bad), its about moving it from `~/.electron-cash` to `~/.config/electron-cash` with backwards compatibility

**Comment by alexander255** [comment]
@fyookball: I mostly agree about your stance on splitting up the files. Basically we're discussing three seperate things here, so maybe I clear things up a bit by outlining each change (and the reasons for doing it) below. From my point of view only (1) and (2) are important, (3) can be done as well but it only makes sense if you feel like this is something you would want to have in EC.

### 1. Moving `~/.electron-cash` to `${XDG_DATA_HOME:-~/.local/share}/electron-cash`

(It really shoud have been `XDG_DATA_HOME` from the start, sorry about this.)

Benefits:

  * Removes clutter in the home directory
  * Allows file indexers, backup tools, etc to easily identify the different *general* types of content
     * File indexers are for instance not interested in any of this data, while backup tools generally want to backup the users configuration (and other program data), but not their app's cache files, …
  * Allows app configuration and user data files to be relocated in managed setups (work, university, …)

Drawbacks:

  * Requires a migration (or using two different save directories like in the current patch)

This is a must-have in the long run.

### 2. Move the `cache` subdirectory to `${XDG_CACHE_HOME}`

Benefits:

  * As with point 1 it helps tools (particularily backup tools in this case) to make sense of your home directory
  * It allows disk cleaners to quickly find out which files make be deleted to reduce space

Drawbacks:

  * None, only a code update would be required – the cached data can be downloaded again on next start

This is a must-have as well.

### 3. Split data between `XDG_CONFIG_HOME` and `XDG_DATA_HOME`

Benefits:

  * Copying the (hopefully small) `~/.config` directory between computers restores most user settings, but may require to reobtaining lots extra data such as installed plugins/apps/extensions/DLCs/whatever, downloaded updates of components 

Drawbacks:

  * Probably lots of works and also requires a migration

This one is quite optional.

**Comment by marceloneil** [comment]
I think 1 and 2 are doable and useful, although my (slightly paranoid) worry is that there are some users that wouldn't have these XDG env variables. Do you know if this is at all possible for posix systems?

**Comment by alexander255** [comment]
@marceloneil: Some users indeed have the given variables not set, but this is not a problem since each XDG directory has a clearly defined default value:

  1. `XDG_CONFIG_HOME` defaults to `~/.config`
  2. `XDG_DATA_HOME` defaults to `~/.local/share`
  3. `XDG_CACHE_HOME` defaults to `~/.cache`

So this problem is solved, if anything the problem more often than not is that people tend to rely on these defaults too much and sometimes forget the actual environment variables.

macOS, as I mentioned in the OP, is different case however as it has its own system for this.

**Comment by marceloneil** [comment]
Oh you're totally right my bad. I've been told that on OSX it still stores the data in `~/.electron-cash` though. I'm wondering if many other apps use `~/.config` on OSX. What do you think of this PR  @cculianu?

**Comment by alexander255** [comment]
(Just to clarify: I do not want to merge this as-is anymore, but gather opionions from the team on how to proceed on this so that the next PR can be better and more usful.)

**Comment by cculianu** [comment]
@marceloneil Yes on OSX it's in ~/.electron-cash.

There are some apps that use ~/.config but the vast majority put stuff in ~/Library/Applicaton Support/<bundle-id> and the like, and ~/Library/Caches/<bundle-id>

I am not sure what I think of this patch... since I haven't kept up with the latest trends in the Linux world.  I suppose it does no harm -- even on OSX.

**Comment by alexander255** [comment]
IMHO, the patch should also move the `user_dir` to `~/Library/Application Support/<bundle-id>` (and `~/Library/Caches/<bundle-id>` for caches) on macOS. This way the correct directory would be used on each platform. It was never my intention to move the stuff to `~/.local/share/electron-cache` on macOS.

I'll propose a new patch so that we can discuss this properly.

**Comment by cculianu** [comment]
Hey Alexander,

That sounds great.  

To add even another wrinkle into the works -- macOS also has this concept of "sandboxed apps" (all mac app store apps are sandboxed) -- this is akin to what happens on iOS.

I am not sure if we want to future-proof for that.

I know the Electrum team is looking into getting Electrum into the mac App store and as such this is something they will have to deal with.

But if we want to future-proof for sandboxing -- it may be recommended that we think about that now.

I THINK the way sandboxing works is the OS creates a namespace/view of the filesystem where ~/Library/Application Support/\<bundle_id\> maps to some deeper directiory (something like ~/Library/Containers/blablabla).  But don't quote me on that. 

The recommended way Apple has is to ask the OS via a 'Foundation' API call for where "Documents" should go and where "Caches" should go and where "Preferences" should go (oh -- by the way -- there are 3 major places for this stuff on OSX rather than 2 on POSIX).

I think in practice though if you write to ~/Library/Application Support/\<bunde id\>, the OS puts it in the right place inside "Containers".  But I am not sure this is guaranteed.

Worst case we could always, for macos, do a 'ctypes' thing where Foundation.framework is loaded on-demand and queried (Foundation probably already sits in memory anyway when Python is running so it may be a cheap op) for the right locations (I can help you with this as I am quite familiar with both ctypes and the macos native api).

Sorry if this is getting too involved -- I just wanted to alert you to the fact that macOS is very bizarre due to Sandboxing.

https://medium.com/macos-app-development/using-app-sandbox-for-macos-app-9bc90556f9ce

**Comment by alexander255** [comment]
@cculianu: I may be able to write a patch incorporating that, but I do not own a Mac to test it. I could test it in my macOS 11 VM, but it's sooo slooow. Could you maybe provide me with the required code to query these functions?
Developing and testing that on my own would (as mentioned) be possible, but quite cumbersome unfortunately.

**Comment by alexander255** [comment]
On a more general note: Is there any way to do a migration of the whole settings directory (similar to a per-wallet migration), that will ask for user consent first?
Since the `SimpleConfig` object is created and accessed so early I'm not sure if that is even feasible, currently. Any ideas on how to do this?

**Comment by cculianu** [comment]
Below is a very long function with a script that essentially returns the "Application Support" and "Caches" directories, plus the BUNDLE identifier.

Tested and works.

I recommend you actually try it out yourself on your macOS.. or you can plug it in and trust it works, and if it fails, maybe fall back to the 'POSIX' scheme?

This requires nothing but 'ctypes' and python 3.5+.. ctypes I think is standard on python installs?

```
#!/usr/bin/env python3

# returns:  ('Application Support' Dir, 'Caches' dir, BUNDLE_NAME) as a tuple.
# Any of these may be 'None' if an error occurred or None can be returned on error
# If they are all valid, one can os.path.join(ret[0],ret[2]) to get a directory path one should put application files in, for example.
def GetMacOSDirsAndBundleName() -> (str, str, str):
    try:
        from ctypes import cdll, c_char_p, util, c_void_p, c_int, create_string_buffer, c_bool, c_long, c_uint
        
        corefoundation = cdll.LoadLibrary(util.find_library('CoreFoundation'))
        
        foundation = cdll.LoadLibrary(util.find_library('Foundation'))
        
        corefoundation.CFStringGetCStringPtr.restype = c_char_p
        corefoundation.CFStringGetCStringPtr.argtypes = [c_void_p, c_int]
        kCFStringEncodingUTF8 = 0x08000100
        
        foundation.NSOpenStepRootDirectory.restype = c_void_p
        foundation.NSOpenStepRootDirectory.argtypes = []
            
        NSApplicationSupportDirectory = 14
        NSCachesDirectory = 13
        NSUserDomainMask = 1
        
        # FOUNDATION_EXPORT NSArray<NSString *> *NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory directory, NSSearchPathDomainMask domainMask, BOOL expandTilde);
        foundation.NSSearchPathForDirectoriesInDomains.restype = c_void_p
        foundation.NSSearchPathForDirectoriesInDomains.argtypes = [c_uint, c_uint, c_bool]
        
        
        corefoundation.CFArrayGetCount.restype = c_long
        corefoundation.CFArrayGetCount.argtypes = [c_void_p]
        
        corefoundation.CFArrayGetValueAtIndex.restype = c_void_p
        corefoundation.CFArrayGetValueAtIndex.argtypes = [c_void_p, c_long]
        
        corefoundation.CFStringGetCString.restype = c_bool
        corefoundation.CFStringGetCString.argtypes = [c_void_p, c_void_p, c_long, c_int]
        
        def CFString2Str(cf : c_void_p) -> str:
            b = corefoundation.CFStringGetCStringPtr(c_void_p(cf), kCFStringEncodingUTF8)
            if b is not None:
                return b.decode('utf-8')
            else:
                buf=create_string_buffer(4096)
                if corefoundation.CFStringGetCString(cfstr, buf, 4096, kCFStringEncodingUTF8):
                    b = bytes(buf)
                    b = b[:b.find(b'\0')]
                    return b.decode('utf-8')
            return None
    
        # Application Support
        anArray = foundation.NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, c_bool(True))
        n = corefoundation.CFArrayGetCount(anArray)
        ret1 = None
        if n > 0:
            cfstr = corefoundation.CFArrayGetValueAtIndex(anArray, 0)
            ret1 = CFString2Str(cfstr)
    
        # Caches
        anArray = foundation.NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, c_bool(True))
        n = corefoundation.CFArrayGetCount(anArray)
        ret2 = None
        if n > 0:
            cfstr = corefoundation.CFArrayGetValueAtIndex(anArray, 0)
            ret2 = CFString2Str(cfstr)
            
        # Bundle Stuff
        corefoundation.CFBundleGetMainBundle.restype = c_void_p
        corefoundation.CFBundleGetMainBundle.argtypes = []
        corefoundation.CFBundleGetIdentifier.restype = c_void_p
        corefoundation.CFBundleGetIdentifier.argtypes = [c_void_p]
        
        ret3 = None
        bundle = corefoundation.CFBundleGetMainBundle()
        if bundle:
            ret3 = CFString2Str(corefoundation.CFBundleGetIdentifier(bundle))
    
        return (ret1, ret2, ret3)
    except:
        import sys
        print("Exception in GetMacOSDirsAndBundleName:", str(sys.exc_info()[1]))
        return None

print(GetMacOSDirsAndBundleName())
```

**Comment by cculianu** [comment]
PS: As for your migration/UI issue  -- yeah I think it's fuzzy because the UI isn't really built until the config object is already established.  I am not sure the best way to handle it.  If it were just me on 1 platform I'd pop up a quick-n-dirty UI to do it.  But I guess we need to play nice with both Kivy and Qt.  No idea.  Maybe others can chime in.. :P

Or.. you can "just do it", set a flag, and have the UI inform the user later that "stuff's been done". With no opt-out. :)

**Comment by cculianu** [comment]
One note about the bundle identifier on macs:  When running under the interpreter, it's going to be different than when the built app is run.  

When the app is "running form source", it'll use the Python interpreter's bundle identifier: org.python.python.

However, when it's the final dist .app file complete with its own "Info.plist" that we ship out, it will be "Electron-Cash" (this is wrong and should really be org.electroncash.Electron-Cash -- but c'est la vie -- the devs that did this initially didn't 'get' macs :) ). 

So we may want to hard-code the bundle identifier in our python sources as always being 'Electron-Cash' as is in the shipped app so that users running from source don't find their wallets missing suddenly and vice-versa. :)

**Comment by alexander255** [comment]
@cculianu: Is that bundle identifier actually valid for the Apple's AppStore? Can it be changed currently without anything breaking? Because as far as I understand the current situation Electron-Cash is distributed as `.app` directory within a DMG disk image and the app's bundle identifier is totally irelevant for this, no?

I'm asking specifically because I'm working towards distributing EC as a Flatpak for Linux (similar concept: each app is sandboxed, etc.) and there each app needs an „app-id“ that looks just like those bundle identifiers you posted. Currently that app-id is *org.electroncash.ElectronCash*, but I'd think it'd be nice if app/bundle/whatever ID could match on all platforms.
Looking at the kivy mobile port, it appears that they use *org.electroncash.ElectronCash3*, so maybe we should use that?

@marceloneil @rt121212121: Do you have any idea how we could migrate the data directory while asking the user first? Just moving it does not strike my like a terribly great idea since the change will be backwards-incompatible.

**Comment by marceloneil** [comment]
@alexander255 Maybe take a look at `storage.py`, I recall when there was a wallet-format upgrade, it would notify the user if they wanted to upgrade. Maybe we could doe something similar? Also, the kivy mobile port is falling apart a bit, the only reason it's _org.electroncash.ElectronCash3_ is because jonald lost his signing key. I think the app-id is fine for right now

**Comment by cculianu** [comment]
@alexander255 :

> Because as far as I understand the current situation Electron-Cash is distributed as .app directory within a DMG disk image and the app's bundle identifier is totally irelevant for this, no?

Actually no, once you create a .app, it has inside it an Info.plist, which is basically an XML document that tells the OS's program launcher some stuff about the app, and amongst the things in that Info.plist, is the app bundle id, which, if you do Cocoa/CoreFoundation/Foundation calls to the OS (largely we don't, except in custom code like this), the bundle ID and stuff in the Info.plist becomes relevant.

OSX has a dual-personality. At it's lowest level it's a 100% POSIX Darwin BSD OS, but it has the extra layer on top of it which is the rest of the system + the GUI which of course has stuff like .app bundles and Info.plists and bundle id's.

As far as the bundle ID itself is concerned -- currently it's legal to "just run" the app but is not a legal one for the app store.  I think they require you to use a reverse domain name.  You can actually make up whatever domain name you want, but it has to be unique and can't steal someone else's trademark.  So you couldn't do com.micro-s0ft or something, but you could do com.BobsApps.ElectronCash.

> Currently that app-id is org.electroncash.ElectronCash, but I'd think it'd be nice if app/bundle/whatever ID could match on all platforms.

I agree with this 10000%.  All we need to do for mac is modify the Info.plis that's auto-generated -- it's a 1 line change in contrib/build-osx/osx.spec:

```

app = BUNDLE(exe,
             version = VERSION,
             name=PACKAGE + '.app',
             icon=electrum+ICONS_FILE,
             bundle_identifier=None,   # <--- this needs to change to com.electroncash.ElectronCash
```

So yeah -- all for this bundle-id-unification.

And glad Linux is going along these lines -- I think sandboxing makes sense for lots of apps from "untrusted" sources to contain the damage/viruses/spying they can do.

If you get this stuff regularized we can even try and distribute Electron Cash through the mac app store! :)

I can help with this stuff as soon as I'm done releasing ElectronCash on iOS. Which should be after July 1st.

**Comment by alexander255** [comment]
@marceloneil: Thanks for the input on the app-id issue!
Regarding the config migration: The main issue here is that the configuration directory will have been long read and in use, by the time we start the GUI (not least because the GUI is selected because of the config), so just asking for migration before use like `storage.py` is pretty much off the table unfortunately.

I see four possible solutions here (in order of preference):

  1. Copy the files on startup (instead of moving them) and then tell the user about this – maybe also offer to delete the old directory
  2. Do a silent migration and then inform the user (like @cculianu suggested)
  3. Pop up a native dialog on startup, *before* reading the config, using something like [`zenity`](https://en.wikipedia.org/wiki/Zenity).
  4. Keep using the old directory directory (like in the current patch, we'd end up having two possible save locations)

Which one do you (and @cculianu too!) believe is best of these options? I'm open to other suggests as weel of course.

@cculianu: Thanks for the detailed analysis! I totally don't mind reading over 2 pages of text, if it helps me get a more solid background on how stuff works!

> I can help with this stuff in a practical sense as soon as I'm done releasing ElectronCash on iOS. Which should be after July 1st (as in: I can do the above modifications myself and test them to make sure they are 100% kosher).

This likely won't land before that time anyways, so don't be too hard on yourself. :slightly_smiling_face: 

But maybe could you just test whether the bundle-ID change itself has unintended sideeffects and that updating that file and that running `./contrib/build-osx/make_osx` is all that is required? Doing so should be rather quick and the more advanced stuff (like the code you posted above) follow in due time from my side. I just really cannot test this part unfortunately.

> As an aside, I'm glad Linux is going along these lines -- I think sandboxing makes sense for lots of apps from "untrusted" sources to contain the damage/viruses/spying they can do.

Me too! That and ease of installation across distros is why I'm pushing for this. Truth to be told through, the current sandbox has quite a lot holes currently and apps can ask for just about anything using sandbox exceptions, but it's a start. What this means is that the sandbox is currently effectively opt-in and apps do have to play nice.
Often times extra sandbox permissions are required because of adaptions needed for apps to run in such an environment not being impleted, but often times it's also that the technology itself to do some things from a sandboxed environment is simply not available.
For instance EC currently has `"--device=all"` (access to all device nodes in `/dev`) and `"--filesystem=/sys/devices:ro"` and `"--filesystem=/sys/class/video4linux:ro"` (access to system information about most devices on the system) in its sandbox exemption list because there is no way to record video from within the sandbox currently and so we need to go back to raw device access for this. We also only require video access for the QR scanner component, but cannot limit it to that either, ofc. The main developer of Flatpak told me to wait for a software named PipeWire that doesn't really exist yet…
But we're getting there I think, so I'm happy to help out with a port – even if it ends being just for easier installation.

Sorry for the blurb about Flatpak; I just wanted to share some experiences. :slightly_smiling_face:

**Comment by marceloneil** [comment]
I had another idea! What if there was a popup at the beginning like the storage.py one, and allowing a migration would add a key to the config, or an env variable or something like that and restart the app. 
I'm not sure how to restart the app with context, but it might be possible?

**Comment by cculianu** [comment]
My vote is seriously to move them and tell them about it after the fact. It's the path of least resistence and 80% of users wouldn't care, most likely.

Also be careful not to lose their files when you code it up (as in, make doubly sure they really got copied and if not, give up).

Those are my 2 cents.  Keep it simple. :)

TBH I like that the files are now in ~/.electron-cash because I'm a dev and I can easily get there.  If they end up living in ~/Library/Application Support/bla bla bla it'll just be more typing for me. Ha! :)

The only caveat with moving them is if users 'downgrade' back to a previous version, then of course the previous version won't see their files.  

Anyway yeah  that bundle-id change in ./contrib/build-osx/make_osx is all that's needed on mac to get a sane bundle id -- I'm 99.9% sure of that without even testing it (I have been doing this a while  -- building for Mac and wrestling with it so i'm fairly certain).

**Comment by alexander255** [comment]
*Reconsidering all options again (since I now have time to actually work on this)…*

 * @cculianu: Moving EC's data files (including their money/wallets!) without user interaction is a no-go IMHO. Yes, 80% won't and 20% will (rightfully) be pissed – for no good reason. The more I consider this, the worse it sounds, so that option is off the table from my side.
 * The popup before startup thing is obviously not going to happen, since nobody wants to maintain an entire second UI implementation for this – even if it would be quite minimal.
 * Retaining the old save directory has basically been rejected (at least nobody expressed any desire in having this).
 * That leaves:
    1. Copying the wallet, leaving the old one in tact, then (later) display an UI popup about the migration being complete and the old directory now being dormant
    2. Using the old wallet and having the UI display a message that a migration is required; if the user agrees set a special flag somewhere and restart causing the migration to happen at that point (@marceloneil's suggestion)

I don't really care which of these two we choose, but here's some extra …

**ad i)** I just updated GnuCash and here's the dialog it poped at me:
GnuCash after 3.0 update
This is basically a 1:1 equivallent of what I suggested doing for EC. (It's always nice to have visuals so I thought it might be good to add this here.)

**ad ii)** In response to @marceloneil: Restarting the app with the same parameters is pretty trivial in EC's case since there are exactly 3 possible kinds of startup inputs that *may* affect this app:

  1. Environment variables (will be retained automatically on `exec`)
  2. Command-line parameters (easy to pass to the new process – just pull them from `sys.argv` and pass them on)
  3. Non-`CLOEXEC` file descriptors (not used by EC, except for `stdout`/`stderr` – also retained automatically during `exec`)

… just clarifying that there is no technical reason no to do this.

.

I honestly don't care which of these we choose, but I'd like have a final decision before I go about implementing this. @marceloneil or whoever, just tell me which of these two you prefer and I'll get to work.

**Comment by cculianu** [comment]
Hey  I love the visuals.  Throw a picture into a comment all of a sudden it's more fun to read.  I'm convinced.  If GNUcash does it that way, we can too.

My vote: I'd be in favor of the simpler solution -- ie the one that GNUcash does 100%.  (Just going ahead and using the new dir, keeping the old dir, and informing the user about it with suggestions on why they may want to delete or keep old dir, as gnucash did).

I think @marceloneil and maybe @fyookball are probably the final says on this.  I live more in my iOS -implementation world.

As an aside -- I can think of *one* reason why restarting the app is a technical challenge going-forward (but this maybe a non-issue from now until the rest of time!): *macOS sandboxing*!

On macOS, if you want to 'sandbox' your app (which we don't currently do) and list it in the mac App Store, you cannot have it restart itself.  The exec call fails spectacularly.  It's lame.. I know!  Not sure why they did that.  There are ways to work around that (using a "helper" app) -- but just as an FYI that if you ever want to sandbox an app -- it can't easily restart itself.

Otherwise yes -- on all platforms Qt runs (with the exception of macOS sandboxe), electron-cash can safely exec() itself.

**Comment by marceloneil** [comment]
I'm also in favour of the GNUcash method - it seems very clean and simple. I'm also a fan that it chooses to copy instead of move, so we could warn the user if something went wrong but still maintain functionality

**Comment by alexander255** [comment]
I opened a new (in-progress) PR that implements the copy-then-notify suggestion: #780 
Currently only the Linux data directories are supported, but I updated both the Qt and the Kivy GUI. If you want to take a look at the currently implementation please go ahead as it won't change much after this, otherwise I'll write here once I'm done with writing this.

**Comment by alexander255** [comment]
Closing this in favour of #780.
